### ISSUE TITLE
Update request: emscripten 3.1.51 → 3.1.57

### ISSUE THREAD

**Issue by ajbt200128** [NixOS/nixpkgs#306649]
- Package name: emscripten
- Latest released version: 3.1.57

- Current version on the unstable channel: 3.1.51
- Current version on the stable/release channel: 3.1.47

- [X] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**
@qknight @matthewbauer @raitobezarius @willcohen 

-----

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by tracteurblinde** [comment]
tl;dr: 3.1.57 is not currently compatible with the version of llvm in tree (18.1.5). The latest compatible version with llvm 18.1.5 appears to be Emscripten 3.1.54.

The emscripten nixpkg (3.1.51) currently uses an LLVM17 environment which needs be migrated to LLVM18. The configuration as it exists today doesn't work in some cases. (Godot4.3dev6 web template builds failed missing `--keep-section=target_features` in `wasm-ld`, prompting finding this issue)

Emscripten 3.1.51 seems to be compatible with the LLVM 18.1.5 currently in nixpkgs unstable. Emscripten 3.1.59 (current as of this comment) and 3.1.57 are not compatible with that version of LLVM (In my usage, ran into errors with `--no-growable-memory` and `--initial-memory`).

These flags were introduced in Emscripten 3.1.55 making the last compatible version against the current in-tree dependencies 3.1.54.

<details><summary>This override was able to successfully build the godot4.3dev6 web templates using Emscripten 3.1.54</summary>
<p>

```nix
# nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable/3281bec7174f679eabf584591e75979a258d8c40";
  emscripten = pkgs.emscripten.overrideAttrs (old: rec {
    version = "3.1.54";
    src = fetchFromGitHub {
      owner = "emscripten-core";
      repo = "emscripten";
      rev = version;
      sha256 = "IY57pFgNGxC2tpzlEhn/ZFZgmL4spCJ1ZnXQID9lrT8=";
    };
    llvmEnv = pkgs.symlinkJoin {
      name = "emscripten-llvm-${version}";
      paths = with pkgs.llvmPackages_18; [ clang-unwrapped clang-unwrapped.lib lld llvm ];
    };
    patches = [
      (pkgs.substituteAll {
        src = "${nixpkgs}/pkgs/development/compilers/emscripten/0001-emulate-clang-sysroot-include-logic.patch";
        resourceDir = "${llvmEnv}/lib/clang/18/";
      })
    ];
    buildPhase = ''
      runHook preBuild

      patchShebangs .

      # emscripten 3.1.54 requires LLVM tip-of-tree instead of LLVM 18
      sed -i -e "s/EXPECTED_LLVM_VERSION = 19/EXPECTED_LLVM_VERSION = 18.1/g" tools/shared.py

      # fixes cmake support
      sed -i -e "s/print \('emcc (Emscript.*\)/sys.stderr.write(\1); sys.stderr.flush()/g" emcc.py

      sed -i "/^def check_sanity/a\\  return" tools/shared.py

      echo "EMSCRIPTEN_ROOT = '$out/share/emscripten'" > .emscripten
      echo "LLVM_ROOT = '${llvmEnv}/bin'" >> .emscripten
      echo "NODE_JS = '${pkgs.nodejs_18}/bin/node'" >> .emscripten
      echo "JS_ENGINES = [NODE_JS]" >> .emscripten
      echo "CLOSURE_COMPILER = ['${pkgs.closurecompiler}/bin/closure-compiler']" >> .emscripten
      echo "JAVA = '${pkgs.jre}/bin/java'" >> .emscripten
      # to make the test(s) below work
      # echo "SPIDERMONKEY_ENGINE = []" >> .emscripten
      echo "BINARYEN_ROOT = '${pkgs.binaryen}'" >> .emscripten

      # make emconfigure/emcmake use the correct (wrapped) binaries
      sed -i "s|^EMCC =.*|EMCC='$out/bin/emcc'|" tools/shared.py
      sed -i "s|^EMXX =.*|EMXX='$out/bin/em++'|" tools/shared.py
      sed -i "s|^EMAR =.*|EMAR='$out/bin/emar'|" tools/shared.py
      sed -i "s|^EMRANLIB =.*|EMRANLIB='$out/bin/emranlib'|" tools/shared.py

      runHook postBuild
    '';
  });
```

</p>
</details>
